monitoring/prometheus/metrics.go

```diff
@@ -0,0 +1,251 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package prometheus provides a Prometheus-based implementation of the
+// MetricFactory abstraction.
+package prometheus
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/monitoring"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+// MetricFactory allows the creation of Prometheus-based metrics.
+type MetricFactory struct {
+	Prefix string
+}
+
+// NewCounter creates a new Counter object backed by Prometheus.
+func (pmf MetricFactory) NewCounter(name, help string, labelNames []string) monitoring.Counter {
+	if labelNames == nil || len(labelNames) == 0 {
+		counter := prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: pmf.Prefix + name,
+				Help: help,
+			})
+		prometheus.MustRegister(counter)
+		return &Counter{single: counter}
+	}
+
+	vec := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: pmf.Prefix + name,
+			Help: help,
+		},
+		labelNames)
+	prometheus.MustRegister(vec)
+	return &Counter{labelNames: labelNames, vec: vec}
+}
+
+// NewGauge creates a new Gauge object backed by Prometheus.
+func (pmf MetricFactory) NewGauge(name, help string, labelNames []string) monitoring.Gauge {
+	if labelNames == nil || len(labelNames) == 0 {
+		gauge := prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name: pmf.Prefix + name,
+				Help: help,
+			})
+		prometheus.MustRegister(gauge)
+		return &Gauge{single: gauge}
+	}
+	vec := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: pmf.Prefix + name,
+			Help: help,
+		},
+		labelNames)
+	prometheus.MustRegister(vec)
+	return &Gauge{labelNames: labelNames, vec: vec}
+}
+
+// NewHistogram creates a new Histogram object backed by Prometheus.
+func (pmf MetricFactory) NewHistogram(name, help string, labelNames []string) monitoring.Histogram {
+	if labelNames == nil || len(labelNames) == 0 {
+		histogram := prometheus.NewHistogram(
+			prometheus.HistogramOpts{
+				Name: pmf.Prefix + name,
+				Help: help,
+			})
+		prometheus.MustRegister(histogram)
+		return &Histogram{single: histogram}
+	}
+	vec := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: pmf.Prefix + name,
+			Help: help,
+		},
+		labelNames)
+	prometheus.MustRegister(vec)
+	return &Histogram{labelNames: labelNames, vec: vec}
+}
+
+// Counter is a wrapper around a Prometheus Counter or CounterVec object.
+type Counter struct {
+	labelNames []string
+	single     prometheus.Counter
```

Comment:
Yep, it's an [interface](https://godoc.org/github.com/prometheus/client_golang/prometheus#Counter)